File: .claude/skills/re-review-delta/scripts/conditions.py
"""Reconstruct the conditions ledger from posted review comments.

Deterministic parse of conventional-comment review threads (see
config/comment-vocab.md) into the PRRecord `conditions` shape. This is the
"reconstructed each run, not stored" half (trap #4). Status stays `open` here —
verifying met/not_met/moot against the diff is the resident's job
(re-review-delta SKILL.md), because GitHub "resolved" is only a claim (trap #3).
"""

from __future__ import annotations

import re
from typing import Any

# `label(decoration): subject`, tolerant of leading markdown (**, _, >, spaces).
_LABEL = re.compile(
    r"^[\s>*_~`]*"
    r"(?P<label>issue|suggestion|question|nitpick|praise|todo|thought)"
    r"\s*(?:\(\s*(?P<decoration>[a-z -]+?)\s*\))?"
    r"[\s*_]*:",
    re.IGNORECASE,
)

# Labels that produce a ledger entry (binding or tracked). Others (question,
# nitpick, praise, todo, thought) are not conditions.
_CONDITION_LABELS = {"issue", "suggestion"}


def parse_label(body: str) -> tuple[str, str | None] | None:
    """Return (label, decoration) from a comment body, or None if unlabeled."""
    if not body:
        return None
    first = body.strip().splitlines()[0] if body.strip() else ""
    m = _LABEL.match(first)
    if not m:
        return None
    decoration = m.group("decoration")
    return (m.group("label").lower(), decoration.lower().strip() if decoration else None)


def classify_kind(label: str, decoration: str | None) -> str | None:
    """Map a conventional label to a PRRecord condition kind, or None."""
    if label == "suggestion":
        return "suggestion"
    if label == "issue":
        if decoration == "non-blocking":
            return "non_blocking"
        # blocking, or bare `issue:` -> treat as blocking (conservative).
        return "blocking"
    return None


def _subject(body: str) -> str:
    first = body.strip().splitlines()[0] if body.strip() else ""
    m = _LABEL.match(first)
    text = first[m.end():].strip() if m else first
    return text


def reconstruct(threads: list[dict], viewer: str) -> list[dict[str, Any]]:
    """Build the conditions ledger from review threads authored by `viewer`."""
    ledger: list[dict[str, Any]] = []
    for thread in threads:
        comments = (thread.get("comments") or {}).get("nodes") or []
        if not comments:
            continue
        root = comments[0]
        author = (root.get("author") or {}).get("login")
        if author != viewer:
            continue
        parsed = parse_label(root.get("body") or "")
        if parsed is None:
            continue
        label, decoration = parsed
        if label not in _CONDITION_LABELS:
            continue
        kind = classify_kind(label, decoration)
        if kind is None:
            continue
        line = thread.get("line") or thread.get("originalLine")
        path = thread.get("path")
        ledger.append({
            "id": thread.get("id"),
            "text": _subject(root.get("body") or ""),
            "kind": kind,
            "status": "open",          # resident sets met/not_met/moot
            "evidence_ref": None,      # resident fills with verification evidence
            "location": f"{path}:{line}" if path else None,
            "author_resolved": bool(thread.get("isResolved")),
            "is_outdated": bool(thread.get("isOutdated")),
        })
    return ledger

File: internal/agent/conditions.go
package agent

import (
	"regexp"
	"strconv"
	"strings"

	"github.com/lancedb/pr-residents/internal/gh"
)

// Condition is one reconstructed review condition. Status starts "open"; the
// resident sets met/not_met/moot by verifying against the delta (trap #3:
// GitHub "resolved" is a claim, not a fact).
type Condition struct {
	ID             string `json:"id"`
	Text           string `json:"text"`
	Kind           string `json:"kind"` // blocking | non_blocking | suggestion
	Status         string `json:"status"`
	EvidenceRef    string `json:"evidence_ref"`
	Location       string `json:"location"`
	AuthorResolved bool   `json:"author_resolved"`
	IsOutdated     bool   `json:"is_outdated"`
}

// labelRE matches `label(decoration): subject`, tolerant of leading markdown.
var labelRE = regexp.MustCompile(
	`^[\s>*_~` + "`" + `]*(issue|suggestion|question|nitpick|praise|todo|thought)\s*(?:\(\s*([a-z -]+?)\s*\))?[\s*_]*:`)

var conditionLabels = map[string]bool{"issue": true, "suggestion": true}

// parseLabel returns (label, decoration) from a comment body, or ok=false.
func parseLabel(body string) (label, decoration string, ok bool) {
	body = strings.TrimSpace(body)
	if body == "" {
		return "", "", false
	}
	first := strings.ToLower(firstLine(body))
	m := labelRE.FindStringSubmatch(first)
	if m == nil {
		return "", "", false
	}
	return m[1], strings.TrimSpace(m[2]), true
}

// classifyKind maps a conventional label to a PRRecord condition kind, or "".
func classifyKind(label, decoration string) string {
	switch label {
	case "suggestion":
		return "suggestion"
	case "issue":
		if decoration == "non-blocking" {
			return "non_blocking"
		}
		return "blocking" // blocking, or bare `issue:` -> conservative
	}
	return ""
}

func firstLine(s string) string {
	if i := strings.IndexByte(s, '\n'); i >= 0 {
		return s[:i]
	}
	return s
}

// subject returns the comment's text after the label prefix.
func subject(body string) string {
	first := firstLine(strings.TrimSpace(body))
	if loc := labelRE.FindStringIndex(strings.ToLower(first)); loc != nil {
		return strings.TrimSpace(first[loc[1]:])
	}
	return first
}

// ReconstructConditions builds the conditions ledger from review threads whose
// root comment is authored by viewer (trap #4: reconstructed each run, never
// stored).
func ReconstructConditions(threads []gh.ReviewThread, viewer string) []Condition {
	ledger := []Condition{}
	for _, t := range threads {
		if t.RootAuthor != viewer {
			continue
		}
		label, decoration, ok := parseLabel(t.RootBody)
		if !ok || !conditionLabels[label] {
			continue
		}
		kind := classifyKind(label, decoration)
		if kind == "" {
			continue
		}
		line := t.Line
		if line == 0 {
			line = t.OriginalLine
		}
		var location string
		if t.Path != "" {
			location = t.Path + ":" + strconv.Itoa(line)
		}
		ledger = append(ledger, Condition{
			ID: t.ID, Text: subject(t.RootBody), Kind: kind, Status: "open",
			Location: location, AuthorResolved: t.IsResolved, IsOutdated: t.IsOutdated,
		})
	}
	return ledger
}
